src/v/kafka/server/audit_log_manager.cc

```diff
@@ -0,0 +1,569 @@
+    }
+    co_await std::move(f);
+}
+
+ss::future<> audit_client::inform(model::node_id id) {
+    vlog(klog.trace, "inform: {}", id);
+
+    // Inform a particular node
+    if (id != client::unknown_node_id) {
+        return do_inform(id);
+    }
+
+    // Inform all nodes
+    return seastar::parallel_for_each(
+      _controller->get_members_table().local().node_ids(),
+      [this](model::node_id id) { return do_inform(id); });
+}
+
+ss::future<> audit_client::do_inform(model::node_id id) {
+    auto& fe = _controller->get_ephemeral_credential_frontend().local();
+    auto ec = co_await fe.inform(id, audit_principal);
+    vlog(klog.info, "Informed: broker: {}, ec: {}", id, ec);
+}
+
+ss::future<> audit_client::create_internal_topic() {
+    constexpr std::string_view retain_forever = "-1";
+    constexpr std::string_view seven_days = "604800000";
+    creatable_topic audit_topic{
+      .name = model::kafka_audit_logging_topic,
+      .num_partitions = config::shard_local_cfg().audit_log_num_partitions(),
+      .replication_factor
+      = config::shard_local_cfg().audit_log_replication_factor(),
+      .assignments = {},
+      .configs = {
+        createable_topic_config{
+          .name = ss::sstring(topic_property_retention_bytes),
+          .value{retain_forever}},
+        createable_topic_config{
+          .name = ss::sstring(topic_property_retention_duration),
+          .value{seven_days}}}};
+    vlog(klog.info, "Creating audit log topic with settings: {}", audit_topic);
+    const auto resp = co_await _client.create_topic({std::move(audit_topic)});
+    if (resp.data.topics.size() != 1) {
+        throw std::runtime_error("Unexpected create topics response");
+    }
+    const auto& topic = resp.data.topics[0];
+    if (topic.error_code == error_code::none) {
+        vlog(klog.debug, "Auditing: created audit log topic: {}", topic);
+    } else if (topic.error_code == error_code::topic_already_exists) {
+        vlog(klog.debug, "Auditing: topic already exists");
+        co_await _client.update_metadata();
+    } else {
+        if (topic.error_code == error_code::invalid_replication_factor) {
+            vlog(
+              klog.warn,
+              "Auditing: invalid replication factor on audit topic, check "
+              "settings, then toggle 'auditing_enabled' after they are "
+              "modified");
+        }
+        const auto msg = topic.error_message.has_value() ? *topic.error_message
+                                                         : "<no_err_msg>";
+        throw internal_topic_creation_failure(
+          fmt::format("{} - error_code: {}", msg, topic.error_code));
+    }
+}
+
+ss::future<> audit_client::shutdown() {
+    _as.request_abort();
+    _send_sem.broken();
+    co_await _gate.close();
+    co_await _client.stop();
+}
+
+ss::future<>
+audit_client::produce(std::vector<client::record_essence> records) {
+    /// TODO: Produce with acks=1, atm -1 is hardcoded into client
```

Comment:
I've noticed a bunch of TODOs.  Are you tracking these anywhere?

Reply:
No just in this source file